### PR TITLE
Use ObjectStore tests for FuturizedStore

### PR DESCRIPTION
**Pull request by aclamk**
This PR intends to provide a way to reuse functional test made for ObjectStore `ceph_test_objectstore` to be applied for any FuturizedStore.
The gist of it is that new Objectstore `xstore` that is implemented using FuturizedStore interface.
On creation `xstore` gets name of FuturizedStore to instantiate. 
Reactor is created for a lifetime of the futurized store.

`ceph_test_objectstore` is extended with ObjectStore index 4 = xstore-objectstore.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/alienstore/alien_store.cc [hunk 1]
```diff
@@ -76,14 +76,15 @@ AlienStore::AlienStore(const std::string& type,
   : type(type),
     path{path}
 {
-  cct = std::make_unique<CephContext>(CEPH_ENTITY_TYPE_OSD);
-  g_ceph_context = cct.get();
-  cct->_conf.set_config_values(values);
```
Comment (ronen-fr) [inline src/crimson/os/alienstore/alien_store.cc:81]: isn't this line required anymore?

File: src/crimson/os/alienstore/alien_store.cc [hunk 2]
```diff
@@ -124,7 +125,12 @@ seastar::future<> AlienStore::stop()
   });
 }
 
-AlienStore::~AlienStore() = default;
+AlienStore::~AlienStore()
+{
+  if (g_ceph_context == cct.get()) {
```
Comment (ronen-fr) [inline src/crimson/os/alienstore/alien_store.cc:130]: am I missing something? cct has a path that does not initialize it

File: src/crimson/os/alienstore/alien_store.cc [hunk 3]
```diff
@@ -76,14 +76,15 @@ AlienStore::AlienStore(const std::string& type,
   : type(type),
     path{path}
 {
-  cct = std::make_unique<CephContext>(CEPH_ENTITY_TYPE_OSD);
-  g_ceph_context = cct.get();
-  cct->_conf.set_config_values(values);
+  if (!g_ceph_context) {
+    cct = std::make_unique<CephContext>(CEPH_ENTITY_TYPE_OSD);
+    g_ceph_context = cct.get();
+  }
```
Comment (ronen-fr) [inline src/crimson/os/alienstore/alien_store.cc:82]: } else {
  cct = g_ceph_context.cct;
}

?

File: src/global/CMakeLists.txt [hunk 4]
```diff
@@ -17,3 +17,8 @@ target_link_libraries(global-static common)
 add_library(global STATIC
   $<TARGET_OBJECTS:libglobal_objs>)
 target_link_libraries(global ceph-common ${EXTRALIBS})
+
+#global that does not include ceph-common
+add_library(global-raw STATIC
```
Comment (ronen-fr) [inline src/global/CMakeLists.txt:22]: Suggest to edit the commit message to add a reason for having that target (not just what it is)

File: src/test/objectstore/CMakeLists.txt [hunk 5]
```diff
@@ -7,10 +7,12 @@ install(TARGETS ceph_perf_objectstore
 add_library(store_test_fixture OBJECT store_test_fixture.cc)
```
Comment (ronen-fr) [inline src/test/objectstore/CMakeLists.txt:7]: commit message phrasing

File: src/crimson/os/alienstore/alien_store.cc [hunk 6]
```diff
@@ -202,7 +202,7 @@ AlienStore::list_objects(CollectionRef ch,
       [=, &objects, &next] {
       auto c = static_cast<AlienCollection*>(ch.get());
       return store->collection_list(c->collection, start, end,
-                                    store->get_ideal_list_max(),
+                                    limit,
```
Comment (ronen-fr) [inline src/crimson/os/alienstore/alien_store.cc:205]: commit msg: 'than'

File: src/crimson/os/alienstore/alien_store.cc [hunk 7]
```diff
@@ -202,7 +202,7 @@ AlienStore::list_objects(CollectionRef ch,
       [=, &objects, &next] {
       auto c = static_cast<AlienCollection*>(ch.get());
```
Comment (ronen-fr) [inline src/crimson/os/alienstore/alien_store.cc:203]: shouldn't it be the min of 'limit' and the conf?

File: src/crimson/os/alienstore/alien_store.cc [hunk 8]
```diff
@@ -194,7 +194,10 @@ AlienStore::list_objects(CollectionRef ch,
   assert(tp);
   return do_with_op_gate(std::vector<ghobject_t>(), ghobject_t(),
                          [=] (auto &objects, auto &next) {
-    objects.reserve(limit);
+    if (limit <= 100) {
```
Comment (ronen-fr) [inline src/crimson/os/alienstore/alien_store.cc:197]: Here too - I would reserve to min of request and, let's say, 100.

File: src/os/bluestore/BlueStore.cc [hunk 9]
```diff
@@ -12083,7 +12083,9 @@ int BlueStore::omap_get_values(
       r = -ENOENT;
       goto out;
     }
-    iter->upper_bound(*start_after);
```
Comment (ronen-fr) [inline src/os/bluestore/BlueStore.cc:12085]: not sure if related to this PR?

File: src/os/bluestore/BlueStore.cc [hunk 10]
```diff
@@ -4617,6 +4617,7 @@ BlueStore::~BlueStore()
   ceph_assert(bluefs == NULL);
```
Comment (ronen-fr) [inline src/os/bluestore/BlueStore.cc:4617]: 'becomes'

File: src/crimson/os/alienstore/alien_store.cc [hunk 11]
```diff
@@ -159,8 +159,10 @@ seastar::future<> AlienStore::umount()
     return tp->submit([this] {
       return store->umount();
     });
-  }).then([] (int r) {
+  }).then([this] (int r) {
     assert(r == 0);
+    // revive op_gate in case we wound like to mount again
```
Comment (ronen-fr) [inline src/crimson/os/alienstore/alien_store.cc:164]: would

File: src/os/bluestore/BlueStore.cc (hunk 9 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:12085]: Well, it is related to this PR only accidentally.
It is an obvious bug but was not exposed until this PR.

File: src/crimson/os/alienstore/alien_store.cc (hunk 1 above)
Comment (aclamk) [reply src/crimson/os/alienstore/alien_store.cc:81]: I reworded commit message. Please check if it now makes sense.

**Comment by aclamk** [comment]
This PR requires #46507 and #46509 fixes.

**Comment by athanatos** [comment]
@aclamk Add me as a reviewer once you want me to have a look.

**Pull request closed without merging**